Prevent GraphArg from keeping real tensors live

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100515

This may potentially fix an OOM on IG Cover model (Meta only: T152238176)

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

ezyang committed:
Prevent GraphArg from keeping real tensors live

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -135,8 +136,16 @@ class GraphArg:
+    def example(self):
+        if isinstance(self._example, TensorWeakRef):
+            return self._example()
+        else:
+            return self._example
```
assert SymInt so that real tensors don't creep back in?

ezyang replied:
I'd probably prefer to unwind the example/fake_example Tensor/SymInt/FakeTensor confusion

yf225 reviewed (approved):
Can confirm that this fixes the OOM error. Thanks!

ezyang committed:
Update on "Prevent GraphArg from keeping real tensors live"


This may potentially fix an OOM on IG Cover model (Meta only: T152238176)

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang committed:
Update on "Prevent GraphArg from keeping real tensors live"


This may potentially fix an OOM on IG Cover model (Meta only: T152238176)

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
